src/tmpfiles/tmpfiles.c

```
@@ -1261,7 +1261,9 @@ static int create_item(Item *i) {
                 log_debug("%s directory \"%s\".", creation_mode_verb_to_string(creation), i->path);
 
                 if (IN_SET(i->type, CREATE_SUBVOLUME_NEW_QUOTA, CREATE_SUBVOLUME_INHERIT_QUOTA)) {
-                        r = btrfs_subvol_auto_qgroup(i->path, 0, i->type == CREATE_SUBVOLUME_NEW_QUOTA);
+                        r = path_check_fstype(i->path, BTRFS_SUPER_MAGIC);
+                        if (r == 1)
+                                 r = btrfs_subvol_auto_qgroup(i->path, 0, i->type == CREATE_SUBVOLUME_NEW_QUOTA);
```

note that a check like this is actually done inside of btrfs_subvol_auto_qgroup() anyway, as we don't want to issue btrfs ioctls on fds that are not btrfs, since ioctls are after all not assigned uniquely. The only difference appears to be that btrfs_subvol_auto_qgroup() operates on a normal fd, while  path_check_fstype() operates on an O_PATH one...

I wonder if we could adjust  btrfs_subvol_auto_qgroup() to use an O_PATH fd, too? Not sure if the btrfs ioctls would work on an O_PATH fd though...

What are the precise implications of O_PATH on autofs mounts?

Hmm, I am also a bit concerned about the fact that this check and the btrfs operation are two independent steps... This is racy after all, and I'd feel more comfortable if there was a way to somehow do all this with the same fd or so, so that we at least know we are talking about the same file system object the whole time...

---

ioctl's don't work on O_PATH file descriptors.  fstat and fstatfs and a few fcntls() work, and they can be used as the starting point for openat() and friends. Hmm... man page doesn't mention fstatfs. I should fix that.
O_PATH on an autofs mount point will not trigger the mount.  If it has previously been triggered and something is mounted on top of the autofs mount point, then open(O_PATH) will get whatever is on "top".
You could get btrfs_subvol_auto_qgroup() to open with O_PATH, check if it is btrfs, then if it is - openat(fd, ".", O_DIRECTORY|O_RDONLY...) to get an fd for ioctl.  That would avoid many races.... though if some other filesystem was mounted on the directory you could lose.  How much paranoia is useful here?

The reason that I didn't offer code to check everything for autofs is that I think autofs should be as transparent as possible.  In this case, it isn't that we don't want autofs, it is that we *do* want btrfs.  Actually would be be more correct to be testing if "/" is btrfs, and skipping btrfs_subvol_auto_qgroup() in that case?  The documentation for 'v' suggests "/" must be btrfs, and 'q' and 'Q'  are based on 'v'.  Would it then be correct to test if i->path is the same filesystem as "/" ???

---

> O_PATH on an autofs mount point will not trigger the mount.

Ah! good to know. Would be excellent if that could be documented in the man page, too...

> You could get btrfs_subvol_auto_qgroup() to open with O_PATH, check if it is btrfs, then if it is - openat(fd, ".", O_DIRECTORY|O_RDONLY...) to get an fd for ioctl.

Similar to that, couldn't we do an open("/proc/self/fd/%i") to convert an O_PATH fd to a non-O_PATH fd? or is that flag inherited in that case?

> The reason that I didn't offer code to check everything for autofs is that I think autofs should be as transparent as possible. In this case, it isn't that we don't want autofs, it is that we do want btrfs. 

Oh, not quite. Note that "q", "v" and "Q" degrade gracefully to "d" if btrfs isn't used, and the offending lines also carry 0755 as mode to apply to these dirs. Hence, regardless if btrfs is used or not, the mkdir and chmod will trigger any autofs set up this way... Hence yupp, I think the way to avoid autofs deadlocks is to check for autofs here...

So my current thinking would be to build on chase_symlinks() for this (chase_symlinks() is our reeimplementation of glibc's canonicalize_file_name() but can operate relative to some specified root directory). We could add a new flag to it, which would stop chasing as soon as a component that is an autofs mount point. In that case it would return a specific new error code, and tmpfiles.d would look for that before realizing a line and skip the line in that case.

---

I was wrong about openat(fd, ".", ...).  It doesn't access anything mounted on "fd" - it always returns the same object as 'fd'.  open("/proc/self/fd/%i") does the same thing.
So path_set_perms() in tmpfiles.c won't trigger an automount - it will change permissions on the automount point (tested and confirmed)
mkdir_parents() calls is_dir() first.  This will detect a directory (without triggering the automount) and so won't make the directory.  So mkdir/chmod/chown won't trigger the automount.
Enhancing chase_symlinks() to fail on an automount should be easy.  I assume you are suggesting putting the test in parse_line() - only when globs aren't expected.  I have a patch for that and will push it.
It silently ignores any (non-glob) path which would pass through any part of an autofs filesystem.  I realised that not all directories in an autofs filesystem are necessarily automount points, and it is not possible to differentiate between that which are, and those which aren't.